crates/typst/src/layout/align.rs

```diff
@@ -445,6 +476,150 @@ cast! {
+/// `SpecificAlignment<HAlignment, OuterVAlignment>` does not allow vertical alignment
+/// position "center", because `V = OuterVAlignment` doesn't have it.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy,
+    V: Into<VAlignment> + Copy,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    H(H),
+    V(V),
+    Both(H, V),
+}
+
+impl<H, V> SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy,
+    V: Into<VAlignment> + Copy,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    /// The horizontal component.
+    pub const fn x(self) -> Option<H> {
+        match self {
+            Self::H(x) | Self::Both(x, _) => Some(x),
+            Self::V(_) => None,
+        }
+    }
+
+    /// The vertical component.
+    pub const fn y(self) -> Option<V> {
+        match self {
+            Self::V(y) | Self::Both(_, y) => Some(y),
+            Self::H(_) => None,
+        }
+    }
+}
+
+impl Add<OuterVAlignment> for HAlignment {
+    type Output = SpecificAlignment<HAlignment, OuterVAlignment>;
+
+    fn add(self, rhs: OuterVAlignment) -> Self::Output {
+        SpecificAlignment::Both(self, rhs)
+    }
+}
+
+impl Add<HAlignment> for OuterVAlignment {
+    type Output = SpecificAlignment<HAlignment, OuterVAlignment>;
+
+    fn add(self, rhs: HAlignment) -> Self::Output {
+        SpecificAlignment::Both(rhs, self)
+    }
+}
+
+impl<H, V> From<SpecificAlignment<H, V>> for Alignment
+where
+    H: Into<HAlignment> + Copy,
+    V: Into<VAlignment> + Copy,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    fn from(value: SpecificAlignment<H, V>) -> Alignment {
+        type FromType<H, V> = SpecificAlignment<H, V>;
+        match value {
+            FromType::H(x) => Alignment::H(x.into()),
+            FromType::V(y) => Alignment::V(y.into()),
+            FromType::Both(x, y) => Alignment::Both(x.into(), y.into()),
+        }
+    }
+}
+
+impl<H, V> Reflect for SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy + Reflect,
+    V: Into<VAlignment> + Copy + Reflect,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    fn input() -> CastInfo {
+        H::input() + V::input()
+    }
+
+    fn output() -> CastInfo {
+        H::output() + V::output()
+    }
+
+    fn castable(value: &Value) -> bool {
+        H::castable(value) || V::castable(value)
+    }
+}
+
+impl<H, V> IntoValue for SpecificAlignment<H, V>
+where
+    H: Into<HAlignment> + Copy + Reflect,
+    V: Into<VAlignment> + Copy + Reflect,
+    HAlignment: Into<H>,
+    VAlignment: Into<V>,
+{
+    fn into_value(self) -> Value {
+        let alignment = match self {
+            Self::H(x) => Alignment::H(x.into()),
+            Self::V(y) => Alignment::V(y.into()),
+            Self::Both(x, y) => Alignment::Both(x.into(), y.into()),
+        };
+        alignment.into_value()
+    }
+}
+
```

Comment:
No need to repeat the previous impl.

(If the Reflect bounds were really needed, feel free to add them back. But they don't seem to be at a first glance, unless Reflect is a supertrait of IntoValue, which it might very well be, and then you'd need to add the bounds back for the code to compile.)

```suggestion
impl<H, V> IntoValue for SpecificAlignment<H, V>
where
    HAlignment: From<H>,
    VAlignment: From<V>,
{
    fn into_value(self) -> Value {
       Alignment::from(self).into_value()
    }
}

```